src/context.rs

```diff
@@ -128,16 +130,19 @@ impl<'a, T: ArenaObj> Ptr<T> {
     pub fn try_deref_mut(&self, ctx: &'a Context) -> Option<RefMut<'a, T>> {
         T::get_arena(ctx)
             .get(self.idx)
-            .unwrap()
-            .try_borrow_mut()
-            .ok()
+            .and_then(|t| t.try_borrow_mut().ok())
     }
 
     /// Create a unique (to the arena) name based on the arena index.
     pub fn make_name(&self, name_base: &str) -> String {
         let idx = self.idx.into_raw_parts();
         name_base.to_string() + "_" + &idx.0.to_string() + "_" + &idx.1.to_string()
     }
+
+    /// Returns true if pointee is still in the arena, and not erased/deallocated.
+    pub fn is_alive(&self, ctx: &Context) -> bool {
```

Comment:
Hi! I was trying to check if an op is erased from the arena and found nothing, so I added `Ptr::is_alive` (not a fan of this name). I also make `try_deref*` methods to check if the pointee is still in the arena. Besides, I don't like `Ptr::is_linked` crash if the pointee is not in the arena. Do you think we should check there as well?

Here is my progress so far in OmniZK in using pass infra - https://github.com/greenhat/omnizk/blob/add-miden-new-ozkir/crates/ir-transform/src/miden/lowering/cf_lowering.rs

Reply:
@vaivaswatha :point_up:

Reply:
> Hi! I was trying to check if an op is erased from the arena and found nothing, so I added `Ptr::is_alive` (not a fan of this name). I also make `try_deref*` methods to check if the pointee is still in the arena. Besides, I don't like `Ptr::is_linked` crash if the pointee is not in the arena. Do you think we should check there as well?
> 
> Here is my progress so far in OmniZK in using pass infra - https://github.com/greenhat/omnizk/blob/add-miden-new-ozkir/crates/ir-transform/src/miden/lowering/cf_lowering.rs

Sorry I somehow missed this last week. Just saw it after your ping today.

I'm curious, in what situation is `Ptr::is_alive`  useful? In other words, I'm trying to see why having a `Ptr` to an erased entity isn't always a bug. (It's something like a dangling pointer, but safe as it'll panic if dereferenced).

Reply:
I'm using it at https://github.com/vaivaswatha/pliron/pull/4/files#diff-bb35f6399df11ed17d92ad8c06ec15cdeb5bfd973838cdedae6d0b593c1b2df1R238 when I'm iterating over the collected(walk) ops to check if the op was not erased by the pass ran on the previous ops.

Reply:
> I'm using it at https://github.com/vaivaswatha/pliron/pull/4/files#diff-bb35f6399df11ed17d92ad8c06ec15cdeb5bfd973838cdedae6d0b593c1b2df1R238 when I'm iterating over the collected(walk) ops to check if the op was not erased by the pass ran on the previous ops.

Isn't it then the Rewriter's responsibility to keep track of erased `Op`s (since it's notified about it anyway as the erase happens via the rewriter),  and check against that before trying to do something else.

Somehow it seems to me that having dangling pointers is a bad idea and checking if it's a dangling one is just a patch / workaround and not a solution.

Having said that, if this is important for you (ie no other way of doing it otherwise), I'm ok with it.

Reply:
I agree that having a dangling pointer is bad, but I did not find a better solution. The pass is traversing a tree while changing the tree in place. I'd use Rc<RefCell<T>> for a tree defined in Rust types.

The Rewriter only applies and reports changes to the plugged Listener. It could be the OperationConvertor's job to analyze Listener's records on every iteration on the collected ops and sync the remaining(and already processed?) ops in the iterator. It seems error-prone and quite complex.

I don't think we can prevent "dangling" Ptr without some tracking in the Ptr itself and/or in Context.

Reply:
>It could be the `OperationConvertor`'s job 

In this case that seems to be the right way to go. It just needs to have a member field `erased_ops: FxHashSet<Ptr<Operation>>` and update/query that. Wouldn't that work?

Reply:
Good idea! I think this might work. Thanks!
I was almost ready to rip off the listener API and make Rewriter and the whole conversion API simpler, but here is the case for it.

Reply:
Removed `Ptr::is_alive` in https://github.com/vaivaswatha/pliron/pull/4/commits/67c3b89f492e702a8df4be606d2c2af508883e4b

Reply:
> Removed `Ptr::is_alive` in [67c3b89](https://github.com/vaivaswatha/pliron/commit/67c3b89f492e702a8df4be606d2c2af508883e4b)

what about `try_deref`?

Reply:
Do you want it to crash on dangling Ptr? I think it's unexpected for `try_*' to crash.

Reply:
> Do you want it to crash on dangling Ptr?  I think it's unexpected for `try_*' to crash.

I think so, yes. The `try` is for interior mutability (`RefCell`) and not for the pointer validity itself. 

I do see your point though. But until we have a strong reason otherwise, let's just treat dangling pointers as a bug.

Reply:
If I squint enough, I can see that both (interior mutability vs. pointer validity) errors might manifest only in certain conditions depending on the IR that is being processed. That means that on some inputs, my compilation pipeline hard crashes, and in case of pointer validity, I have no way to handle it gracefully.

Reply:
When iterating over the IR during a transform, it's possible that we arrive at an instruction (Operation) the second time (when it's already borrowed mutably). For such algorithms, this is useful. It isn't an error, but is how the algorithm works. Although I admit, an argument can be made that it can be handled by a check similar to the `erased_ops.contains()` one we discussed y'day.

>That means that on some inputs

I wouldn't say that it's a function of the input. If it crashes due to a pointer validity, then it's a compiler bug. It isn't something that you attempt to handle post-access at all. There shouldn't have existed a dangling pointer in the first place.

Do you have a requirement for this now? That `try_deref` not panic on dangling pointers? (I'm assuming that you already added the `erased_op` check in the conversions). If yes, then we can see why that situation arises and take a call. Otherwise, for now let dangling pointers be a panic.

Reply:
> Do you have a requirement for this now? That `try_deref` not panic on dangling pointers? (I'm assuming that you already added the `erased_op` check in the conversions). If yes, then we can see why that situation arises and take a call. 

No. `erased_op` tracking list solved that case. 

> Otherwise, _for now_ let dangling pointers be a panic. We can always add it back if there's a real need for the existence of dangling pointers.

Ok. I will restore the original behavior and use `expect` to explain the error.

Reply:
That sounds fair.

Reply:
Done in https://github.com/vaivaswatha/pliron/pull/4/commits/733255e9d1c36754af4594954750e17ee07a2b55